rust/rope/src/diff.rs

```diff
@@ -0,0 +1,329 @@
+        let mut builder = DiffBuilder::default();
+
+        // before doing anything, scan top down and bottom up for like-ness.
+        let mut scanner = RopeScanner::new(base, target);
+        let (start_offset, diff_end) = scanner.find_min_diff_range();
+        let target_end = target.len() - diff_end;
+
+        if start_offset > 0 {
+            builder.copy(0, 0, start_offset);
+        }
+
+        // if our preliminary scan finds no differences we're done
+        if start_offset == base.len() && target.len() == base.len() {
+            return builder.to_delta(base, target);
+        }
+
+        // TODO: because of how `lines_raw` returns Cows, we can't easily build
+        // the lookup table without allocating. The eventual solution would be
+        // to have a custom iter on the rope that returns suitable chunks.
+        let base_string = String::from(base);
+        let line_hashes = make_line_hashes(&base_string, MIN_SIZE);
+
+        let line_count = target.measure::<LinesMetric>() + 1;
+        let mut matches = Vec::with_capacity(line_count);
+
+        let mut targ_line_offset = 0;
+        let mut prev_base = 0;
+
+        let mut needs_subseq = false;
+        for line in target.lines_raw(start_offset..target_end) {
+            let non_ws = non_ws_offset(&line);
+            if line.len() - non_ws >= MIN_SIZE {
+                if let Some(base_off) = line_hashes.get(&line[non_ws..]) {
+                    let targ_off = targ_line_offset + non_ws;
+                    matches.push((start_offset + targ_off, *base_off));
+                    if *base_off < prev_base {
+                        needs_subseq = true;
+                    }
+                    prev_base = *base_off;
+                }
+            }
+            targ_line_offset += line.len();
+        }
+
+        // we now have an ordered list of matches and their positions.
+        // to ensure that our delta only copies non-decreasing base regions,
+        // we take the longest increasing subsequence.
+        // TODO: a possible optimization here would be to expand matches
+        // to adjacent lines first? this would be at best a small win though..
+
+        let longest_subseq = if needs_subseq {
+            longest_increasing_region_set(&matches)
+        } else {
+            matches
+        };
+
+        // for each matching region, we extend it forwards and backwards.
+        // we keep track of how far forward we extend it each time, to avoid
+        // having a subsequent scan extend backwards over the same region.
+        let mut prev_end = start_offset;
+
+        for (targ_off, base_off) in longest_subseq {
+            if targ_off <= prev_end { continue; }
+            let (left_dist, mut right_dist) = expand_match(base, target, base_off,
+                                                           targ_off, prev_end);
+            if targ_off + right_dist > target_end {
```

Comment:
Can you write this:

    right_dist = right_dist.min(target_end - targ_off)

I _think_ that subexpression is guaranteed nonnegative (thus no integer panic), but haven't verified completely.

Reply:
ah yea that's much cleaner 👍